### ISSUE TITLE
No boxes to NMS!!!!!!!!!!!!!!!!!!!

### ISSUE THREAD

**Issue by WIKI2020** [toandaominh1997/EfficientDet.Pytorch#80]
The code can not be used at all. It can be trained and tested. However, demo.py prompt:"no boxes to NMS", and eval. Py prompt:"AVG map: 0.0"

**Comment by WIKI2020** [comment]
All parameters are set correctly

**Comment by lgl603** [comment]
Me too!

**Comment by Jasper-Bai** [comment]
same issue

**Comment by trulyjupiter** [comment]
+1

**Comment by y78h11b09** [comment]
yeah, there are some problems, run demo.py, still happen no boxes to NMS, but can got mAP 0.5439399966.   and training, the CPU  cost is very hight,  even over 85%

**Comment by midasklr** [comment]
> yeah, there are some problems, run demo.py, still happen no boxes to NMS, but can got mAP 0.5439399966. and training, the CPU cost is very hight, even over 85%

hi,can u infer an image correctly?

**Comment by hawl666** [comment]
@toandaominh1997 can u fix this problem?

**Comment by huwanpeng** [comment]
Has anyone solved it?

**Comment by immarioo** [comment]
same issue.

**Comment by wongtou** [comment]
I had solved this problem by downgrading the version of torchvision from 0.4.2 to 0.4.0

**Comment by immarioo** [comment]
in the readme.md, it says like that:
PyTorch 1.3+
Torchvision 0.4.0+ (We need high version because Torchvision support nms now.)
while the torchvision==0.4.0 should match torch==1.2.0, so which version of you torch？

**Comment by wongtou** [comment]
> in the readme.md, it says like that:
> PyTorch 1.3+
> Torchvision 0.4.0+ (We need high version because Torchvision support nms now.)
> while the torchvision==0.4.0 should match torch==1.2.0, so which version of you torch？

pytorch will downgrade to 1.2.0 meanwhile, but it still work.

**Comment by WIKI2020** [comment]
This EfficientDet Less effective than yolov3. I have verified!

**Comment by immarioo** [comment]
> > in the readme.md, it says like that:
> > PyTorch 1.3+
> > Torchvision 0.4.0+ (We need high version because Torchvision support nms now.)
> > while the torchvision==0.4.0 should match torch==1.2.0, so which version of you torch？
> 
> pytorch will downgrade to 1.2.0 meanwhile, but it still work.
I changed my torch=1.2.0 and torchvision=0.4.0, after trained  10 epoch, its ap is still 0.0 and no boxes  to nms, do you have any ideas?

**Comment by xiaow1998** [comment]
> > > in the readme.md, it says like that:
> > > PyTorch 1.3+
> > > Torchvision 0.4.0+ (We need high version because Torchvision support nms now.)
> > > while the torchvision==0.4.0 should match torch==1.2.0, so which version of you torch？
> > 
> > 
> > pytorch will downgrade to 1.2.0 meanwhile, but it still work.
> > I changed my torch=1.2.0 and torchvision=0.4.0, after trained  10 epoch, its ap is still 0.0 and no boxes  to nms, do you have any ideas?

I wrote a blog: https://www.cnblogs.com/xiaoyh/p/12201934.html, I hope to help you.

**Comment by immarioo** [comment]
thanks, just like the blog you wrote, change this line
parser.add_argument('--network', default='efficientdet-d0', type=str,
                    help='efficientdet-[d0, d1, ..]')
to:
parser.add_argument('--network', default='efficientdet-d1', type=str,
                    help='efficientdet-[d0, d1, ..]')
am i right? btw, i trained my own datasets from scratch.

**Comment by xiaow1998** [comment]
Yes, I train my own data set too

**Comment by WIKI2020** [comment]
只有D1网络可以的话，有毛用

**Comment by CraigWang1** [comment]
Same issue here; I tried downgrading torchvision to 0.4.0 and using efficientdet-d1, but it still shows "No boxes to NMS"

Edit: After further inspection, it seems that during validation, the model actually can predict the boxes, but only a portion of the time. So, at least on my part, it just has to do with training the model more and cleaning up my dataset so that the model can predict better.

Edit2: The model didn't improve even after 150 epochs, so I switched to a different implentation (see below)

**Comment by CraigWang1** [comment]
I've also noticed that it doesn't throw this error the 1st time it calculates map. But, it throws this error the second time, which is very weird.
![1st](https://user-images.githubusercontent.com/55901251/72756511-4b19f100-3b82-11ea-89fe-ec13f78421be.png)

But, it throws this error the second time, which is very weird.
![2nd](https://user-images.githubusercontent.com/55901251/72756526-58cf7680-3b82-11ea-8142-2f80d9d31142.png)

Note: this is during the training

**Comment by quangtn266** [comment]
> Same issue here; I tried downgrading torchvision to 0.4.0 and using efficientdet-d1, but it still shows "No boxes to NMS"
> 
> Edit: After further inspection, it seems that during validation, the model actually can predict the boxes, but only a portion of the time. So, at least on my part, it just has to do with training the model more and cleaning up my dataset so that the model can predict better.

Have you any updated for your result? because I'm also training on d1 model and getting low results for map.

**Comment by CraigWang1** [comment]
@quangtn266 Yes so I tried training more, but even after 150 epochs there was still no improvement. So, I moved over to this different implementation: https://github.com/signatrix/efficientdet

No offense intended for toandaominh, but the one I switched to works insanely well and is much more effective; I didn't have any issues with the other implementation.

Hope this helps

**Comment by tangsipeng** [comment]
pytroch 1.1 and torchvision 0.3.0 . only d0 cant train normaly ,

**Comment by quangtn266** [comment]
> pytroch 1.1 and torchvision 0.3.0 . only d0 cant train normaly ,

you mean that model from d1 to d7 can train normally and only d0 meets the problem, right?

**Comment by tangsipeng** [comment]
@quangtn266  i mean only d0 don't meets the problem .

**Comment by dosemeion** [comment]
I added ‘background’ to the VOC_CLASSES，and this error occurred at the 4th epoch.After deleting the ‘background’, it's back to normal

**Comment by tangsipeng** [comment]
i can train normal now , see the code of  'class EfficientDet' . maybe do test after 10 or more epochs can work well , but at the same time , i found that this Implementation have  lots of difference from the paper.

**Comment by BaiqiangGit** [comment]
Same issue when training d5 for 160 epoches on COCO.

**Comment by CraigWang1** [comment]
The official implementation is out! 
https://github.com/google/automl/tree/master/efficientdet
